### PR TITLE
feat(authentication-service): add asymmetric login in authentication service

### PR DESCRIPTION
**Pull request by Surbhi-sharma1**
## Description
Added asymmetric login in authentication service.

Fixes #GH-1007

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## Checklist:

- [ ] Performed a self-review of my own code
- [x] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/core/src/components/bearer-verifier/providers/services-bearer-asym-token-verifier.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2022 Sourcefuse Technologies
+//
+// This software is released under the MIT License.
+// https://opensource.org/licenses/MIT
+import {inject, Provider} from '@loopback/context';
+import {HttpErrors} from '@loopback/rest';
+import {verify} from 'jsonwebtoken';
+import {VerifyFunction} from 'loopback4-authentication';
+import moment from 'moment-timezone';
```
Comment (akshatdubeysf) [inline packages/core/src/components/bearer-verifier/providers/services-bearer-asym-token-verifier.ts:9]: You should have two verifier providers like this, one symmetric and one asymmetric

File: services/authentication-service/src/component.ts [hunk 2]
```diff
@@ -240,8 +245,28 @@ export class AuthenticationServiceComponent implements Component {
       OauthCodeReaderProvider;
     this.providers[AuthCodeBindings.CODEWRITER_PROVIDER.key] =
       CodeWriterProvider;
-    this.providers[AuthCodeBindings.AUTH_CODE_GENERATOR_PROVIDER.key] =
-      AuthCodeGeneratorProvider;
+
+    if (process.env.JWT_PRIVATE_KEY && process.env.JWT_PRIVATE_KEY !== '') {
+      this.providers[AuthCodeBindings.AUTH_CODE_GENERATOR_PROVIDER.key] =
+        AsymmetricAuthCodeGeneratorProvider;
```
Comment (akshatdubeysf) [inline services/authentication-service/src/component.ts:251]: Do not mix code generator with token provider, be consistent with the naming

File: services/authentication-service/src/component.ts [hunk 3]
```diff
@@ -240,8 +245,28 @@ export class AuthenticationServiceComponent implements Component {
...
+    } else {
+      this.providers[AuthCodeBindings.AUTH_CODE_GENERATOR_PROVIDER.key] =
+        AuthCodeGeneratorProvider;
+    }
+    if (process.env.JWT_PRIVATE_KEY && process.env.JWT_PRIVATE_KEY !== '') {
+      this.providers[AuthCodeBindings.AUTH_TOKEN_GENERATOR_PROVIDER.key] =
+        AsymmetricLoginTokenProvider;
+    } else {
+      this.providers[AuthCodeBindings.AUTH_TOKEN_GENERATOR_PROVIDER.key] =
+        LoginTokenProvider;
```
Comment (akshatdubeysf) [inline services/authentication-service/src/component.ts:261]: This one should be named symmetric token provider

File: services/authentication-service/src/providers/auth-asymmetric-code-provider.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        codePayload.mfa = true;
+        if (
+          this.mfaConfig.secondFactor === STRATEGY.OTP &&
+          this.otpConfig.method === OtpMethodType.OTP
+        ) {
+          await this.otpService.sendOtp(user, client);
+        }
+      }
+
+      const keyValue = fs.readFileSync(process.env.JWT_PRIVATE_KEY ?? '');
```
Comment (akshatdubeysf) [inline services/authentication-service/src/providers/auth-asymmetric-code-provider.ts:50]: This should use just the jwt Token provider, depending on the binding

**Review by akshatdubeysf — CHANGES_REQUESTED** [review]

File: packages/cli/README.md [hunk 5]
```diff
@@ -18,7 +18,7 @@ $ npm install -g @sourceloop/cli
 $ sl COMMAND
 running command...
 $ sl (-v|--version|version)
-@sourceloop/cli/2.4.1 linux-x64 node-v16.13.0
+@sourceloop/cli/2.4.1 linux-x64 node-v16.16.0
```
[the remark this answers is not in the document]
Comment (Surbhi-sharma1) [reply packages/cli/README.md:21]: ok sir

File: packages/core/src/components/bearer-verifier/providers/services-bearer-asym-token-verifier.ts (hunk 1 above)
Comment (Surbhi-sharma1) [reply packages/core/src/components/bearer-verifier/providers/services-bearer-asym-token-verifier.ts:9]: yes sir ,I have provided both ,this one is for asymmetric

File: services/authentication-service/src/component.ts (hunk 2 above)
Comment (Surbhi-sharma1) [reply services/authentication-service/src/component.ts]: ok sir

File: services/authentication-service/src/component.ts (hunk 3 above)
Comment (Surbhi-sharma1) [reply services/authentication-service/src/component.ts]: done

File: services/authentication-service/src/providers/auth-asymmetric-code-provider.ts (hunk 4 above)
Comment (Surbhi-sharma1) [reply services/authentication-service/src/providers/auth-asymmetric-code-provider.ts]: done

File: packages/core/src/components/bearer-verifier/component.ts [hunk 6]
```diff
@@ -29,8 +30,13 @@ export class BearerVerifierComponent implements Component {
     this.models = [RevokedToken];
 
     if (this.config && this.config.type === BearerVerifierType.service) {
-      this.providers[Strategies.Passport.BEARER_TOKEN_VERIFIER.key] =
-        ServicesBearerTokenVerifyProvider;
+      if (process.env.JWT_PUBLIC_KEY && process.env.JWT_PUBLIC_KEY !== '') {
+        this.providers[Strategies.Passport.BEARER_TOKEN_VERIFIER.key] =
+          BearerAsymmetricTokenVerifyProvider;
```
Comment (akshatdubeysf) [inline packages/core/src/components/bearer-verifier/component.ts:35]: Keep the name consistent

File: packages/core/src/components/bearer-verifier/providers/services-bearer-asym-token-verifier.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  implements Provider<VerifyFunction.BearerFn>
+{
+  constructor(@inject(LOGGER.LOGGER_INJECT) public logger: ILogger) {}
+
+  value(): VerifyFunction.BearerFn {
+    return async (token: string) => {
+      let user: IAuthUserWithPermissions;
+
+      try {
+        const publicKey = fs.readFileSync(process.env.JWT_PUBLIC_KEY ?? '');
```
Comment (akshatdubeysf) [inline packages/core/src/components/bearer-verifier/providers/services-bearer-asym-token-verifier.ts:24]: use async await

File: services/authentication-service/src/component.ts [hunk 8]
```diff
@@ -243,6 +248,20 @@ export class AuthenticationServiceComponent implements Component {
...
       AuthCodeGeneratorProvider;
 
+    if (process.env.JWT_PRIVATE_KEY && process.env.JWT_PRIVATE_KEY !== '') {
+      this.providers[AuthCodeBindings.JWT_SIGNER.key] =
+        JWTAsymmetricSignerProvider;
+    } else {
+      this.providers[AuthCodeBindings.JWT_SIGNER.key] =
+        JWTSymmetricSignerProvider;
+    }
+    if (process.env.JWT_PUBLIC_KEY && process.env.JWT_PUBLIC_KEY !== '') {
```
Comment (akshatdubeysf) [inline services/authentication-service/src/component.ts:258]: keep this binding in the same condition as above na?

File: services/authentication-service/src/providers/jwt--asymmetric-signer.provider.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+export class JWTAsymmetricSignerProvider<T>
+  implements Provider<JWTSignerFn<T>>
+{
+  value(): JWTSignerFn<T> {
+    return (data: string | T, options: jwt.SignOptions) => {
+      if (typeof data !== 'string') {
+        data = JSON.stringify(data);
+      }
+      const clientSecret = fs.readFileSync(process.env.JWT_PRIVATE_KEY ?? '');
+      const accessToken = jwt.sign(JSON.parse(data), clientSecret, {
```
Comment (akshatdubeysf) [inline services/authentication-service/src/providers/jwt--asymmetric-signer.provider.ts:15]: is the parse required here?

File: services/notification-service/openapi.md [hunk 10]
```diff
@@ -20,7 +20,7 @@ headingLevel: 2
 
 > Scroll down for code samples, example requests and responses. Select a language for code samples from the tabs above or the mobile navigation menu.
 
-notification microservice.
+notification microservice
```
Comment (akshatdubeysf) [inline services/notification-service/openapi.md:23]: why all the other services are changed?

**Review by akshatdubeysf — CHANGES_REQUESTED** [review]

File: packages/core/src/components/bearer-verifier/component.ts (hunk 6 above)
Comment (Surbhi-sharma1) [reply packages/core/src/components/bearer-verifier/component.ts:35]: It was giving sonar code smell (length greater than 100).

File: services/authentication-service/src/providers/jwt--asymmetric-signer.provider.ts (hunk 9 above)
Comment (Surbhi-sharma1) [reply services/authentication-service/src/providers/jwt--asymmetric-signer.provider.ts:15]: yes Sir ,it is required to parse the data here

File: services/notification-service/openapi.md (hunk 10 above)
Comment (Surbhi-sharma1) [reply services/notification-service/openapi.md]: changes done

File: services/authentication-service/src/component.ts (hunk 8 above)
Comment (Surbhi-sharma1) [reply services/authentication-service/src/component.ts]: done

File: services/authentication-service/src/providers/auth-code-generator.provider.ts [hunk 11]
```diff
@@ -46,13 +54,16 @@ export class AuthCodeGeneratorProvider
...
-          issuer: process.env.JWT_ISSUER,
-          algorithm: 'HS256',
-        }),
+        await this.jwtSigner(
+          codePayload,
+          {
+            expiresIn: client.authCodeExpiration,
+            audience: client.clientId,
+          },
+          client.secret,
```
Comment (akshatdubeysf) [inline services/authentication-service/src/providers/auth-code-generator.provider.ts:65]: But the signed doesn't take third parameter na?

File: services/authentication-service/src/providers/jwt--asymmetric-signer.provider.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+export class JWTAsymmetricSignerProvider<T>
+  implements Provider<JWTSignerFn<T>>
+{
+  value(): JWTSignerFn<T> {
+    return async (data: string | T, options: jwt.SignOptions) => {
+      if (typeof data !== 'string') {
+        data = JSON.stringify(data);
+      }
+      const clientSecret = await fs.readFile(process.env.JWT_PRIVATE_KEY ?? '');
```
Comment (akshatdubeysf) [inline services/authentication-service/src/providers/jwt--asymmetric-signer.provider.ts:14]: This one doesn't take clientSecret?

**Review by akshatdubeysf — CHANGES_REQUESTED** [review]

**Pull request merged**